### PR TITLE
Setup KsatHorFrac

### LINKED ISSUE #228: Include KsatHorFrac estimation method from Awad Ali's MSc thesis
### Kind of request

Adding new functionality

### Enhancement Description

In his MSc thesis, Awad Ali tested Random Forest (RF) and Boosted Regression Trees (BRT) methods to estimate the KsatHorFrac parameter of wflow_sbm. The methods show good results and may be a first step to an alternative for the now uniform KsatHorFrac parameter (with value 100). Awad has already prepared two global maps (one based on RF and one based on the BRT algorithm), which can be found on the wflow P-drive (`hydromt_staging/kstathorfrac_thesis_awadali/`). Because he trained the RF and BRT models on the CAMELS-UK dataset in the UK, the method may not yet be working well in regions with different geology and soil types. But this is a point of further research at the moment.

I think the results are ready for usage in hydromt as an alternative option to the standard uniform KsatHorFrac derivation. In the additional context below, I've provide a function that we use to include one of the KsatHorFrac maps to update an existing wflow_sbm model. Let's see if we can include this in hydromt_wflow in such a way that we can do this up front through a function. Can you assist me in creating a hydromt function that can make use of other KsatHorFrac maps? I am not familiar enough with all functions and procedures in the current version of hydromt to directly open a PR myself, so let's first see where the function belongs and what it should look like. If we make it general enough, you can switch between the 'standard' derivation, the RF and BRT methods (or possibly something else in the future). I can imagine that we update the RF and BRT maps at some point based on future research, but that shouldn't change the hydromt functionality, I assume. 

### Use case

The KsatHorFrac parameter is currently derived as a uniform parameter with value 100, which has to be calibrated later on for improved model performance. Some sort of transfer function for this parameter would be ideal. With Awad's work, we have maps of gridded KsatHorFrac values that can be used in wflow_sbm. These values are based on a ML-based transfer functions between soilgrids data and the KsatHorFrac value. With the current global maps, we expect good results in regions with geology and soil types similar to the UK. For other regions, this has still to be tested.

### Additional Context

Below the function we use to update the wflow model with the new KSatHorFrac maps. The maps are stored at the soilgrids spatial resolution and are store as log10(KSatHorFrac). 

```python
import os
import numpy as np

import hydromt
from hydromt_wflow import WflowModel


# A function adjust_ksathorfrac is defined that takes in a model_folder path. 
# The function adjusts the KsatHorFrac parameter grid based on either a 
# random forest or boosted regression trees based global parameter map.
def adjust_ksathorfrac_and_toml_file(
    model_folder, 
    ksathorfrac_path=None, 
    pred_method=None,
    ):
    """
    Adjust the KsatHorFrac parameter grid with empirical relationship.

    Args:
        model_folder (str): The path to the wflow model.
        ksathorfrac_path (str): optional. the path to the .tif file with the global KsatHorFrac 
                                maps based on the random forest or boosted regression trees
                                methods. Standard set to None.
        pred_method (str): optional. 'RF' for Random Forest, and 'BRT' for Boosted Regression 
                                Trees. If None, the KsatHorFrac value will not be changed. 
                                Standard set to None.

    Returns:
        None.    
    """
    # Open the wflow model
    print(f"Reading wflow model at {model_folder}")
    mod = WflowModel(model_folder, config_fn="wflow_sbm.toml", mode="r+")
    
    # Change KsatHorFrac if requested
    if pred_method is not None:
        if pred_method == "RF" or pred_method == "BRT":
            var_name = pred_method + '_250_KsatHorFrac_global'
        
            # read (clipped by extent) KsatHorFrac with data_catalog.get_rasterdataset
            da_KsatHorFrac = mod.data_catalog.get_rasterdataset(
                data_like = ksathorfrac_path,
                geom = mod.staticgeoms['basins']
                )
            
            # tranform to a logarithmic scale
            da_KsatHorFrac.values = np.log10(da_KsatHorFrac.values)

            # resample KsatHorFrac with grid_from_rasterdataset
            resampled_KsatHorFrac = hydromt.workflows.grid.grid_from_rasterdataset(
                grid_like = mod.staticmaps['KsatHorFrac'],
                ds = da_KsatHorFrac,
                reproject_method = ['average'])
            
            # tranform back the log KsatHorFrac
            resampled_KsatHorFrac[var_name] = 10**resampled_KsatHorFrac[var_name]
            
            # fill missing values
            filled_KsatHorFrac = resampled_KsatHorFrac
            filled_KsatHorFrac[var_name] = filled_KsatHorFrac[var_name].interpolate_na(dim = 'lon', method='linear')

            # fill NA's around the boundary
            filled_KsatHorFrac[var_name] = filled_KsatHorFrac[var_name].interpolate_na(dim = 'lon', method='nearest', fill_value="extrapolate")

            # add and write to staticmap
            mod.set_staticmaps(filled_KsatHorFrac[var_name], 
                            name="KsatHorFrac_"+pred_method)

    # Edit model config
    setting_toml = {}
    if pred_method == "RF":
        setting_toml["input.lateral.subsurface.ksathorfrac"] = "KsatHorFrac_RF"
    elif pred_method == "BRT":
        setting_toml["input.lateral.subsurface.ksathorfrac"] = "KsatHorFrac_BRT"

    # Loop through each setting defined in setting_toml and update 
    # it in the model configuration
    for option in setting_toml:
        mod.set_config(option, setting_toml[option])

    #Write staticmaps and new TOML config
    if pred_method is not None:
        mod.write_staticmaps()
    mod.write_config(config_name="wflow_sbm_calibrated.toml")

    return None
```

### PR DESCRIPTION
**Pull request by dalmijn**
## Issue addressed
Fixes #228 

## Explanation
Made setup method and a workflow function. Added a wflow specific catalog for wflow specific data.

## Checklist
- [x] Updated tests or added new tests
- [x] Branch is up to date with `main`
- [x] Tests & pre-commit hooks pass
- [x] Updated documentation if needed

## Additional Notes (optional)
Added data_catalog should be further discussed. The new ksathorfrac data will be rewritten to a zarr archive.
Data should also be added to the artifact data.

### CODE REVIEW

File: hydromt_wflow/workflows/soilparams.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+    Returns
+    -------
+    _type_
+        _description_
+    """
+    # Transfer data to a logaritmic scale
+    da.values = np.log10(da.values)
+    # Reproject the data
+    da = da.raster.reproject_like(
```
Comment (hboisgon) [inline hydromt_wflow/workflows/soilparams.py:33]: Is there a reason why you are not using the hydromt core method hydromt.workflows.grid.grid_from_rasterdataset for the next steps ? This function includes interpolating, reprojecting and renaming. Even masking if you pass wflow_subcatch.

File: tests/test_model_methods.py [hunk 2]
```diff
@@ -255,6 +255,33 @@ def test_setup_reservoirs(source, tmpdir, example_wflow_model):
 number of reservoirs in model area"
 
 
+def test_setup_ksathorfrac(tmpdir, example_wflow_model):
+    # Read the modeldata
+    model = "wflow"
+    example_wflow_model.read()
+    # Add the wflow datacatalog
+    example_wflow_model.data_catalog.from_yml(
+        "https://raw.githubusercontent.com/Deltares/hydromt_wflow/setup_ksat/data/catalogs/wflow_data.yml"
```
Comment (hboisgon) [inline tests/test_model_methods.py:264]: See the general comment on adding a shortcut for this catalog. For testing, there is no p drive access so I suggest to export the ksathorfrac maps for the Piave extent for testing.
And add one additional test to check if the wflow_data catalog can be read correctly (as we cannot go further and access the data).

**Review by hboisgon — COMMENTED** [review]
Nice changes @dalmijn ! We need to think how to deal with these "predefined catalogs" for wflow so I suggested a few points that we can discuss together.

File: hydromt_wflow/wflow.py [hunk 3]
```diff
@@ -1900,6 +1900,42 @@ def setup_soilmaps(
         ).reset_coords(drop=True)
         self.set_grid(dsout)
 
+    def setup_ksathorfrac(
+        self,
+        ksat_fn: str = "KsatHorFrac_global",
+        ksat_method: str = "BRT_250_KsatHorFrac_global",
+        scale_method: str = "average",
```
[the remark this answers is not in the document]
Comment (dalmijn) [reply hydromt_wflow/wflow.py:1907]: Well, the argument name can be better, but given the fact that are different methods available (*versions of the ksathorfrac estimation), we should probably list them.

File: hydromt_wflow/workflows/soilparams.py (hunk 1 above)
Comment (dalmijn) [reply hydromt_wflow/workflows/soilparams.py:33]: Well, I basically only use the reproject method on the data on the logarithmic scale, after which is immediately goes back to a normal scale. Besides reproject nothing is needed at this stage. Therefore, in my opinion, this gives more clarity to the code.

File: hydromt_wflow/workflows/soilparams.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    scale_method : str
+        _description_
+
+    Returns
+    -------
+    _type_
+        _description_
+    """
+    # Transfer data to a logaritmic scale
+    da.values = np.log10(da.values)
```
[the remark this answers is not in the document]
Comment (dalmijn) [reply hydromt_wflow/workflows/soilparams.py:31]: Good point, we can use log10 from `dask.array` in combination with the `map_blocks` functionality.

File: data/catalogs/wflow_data.yml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  driver: raster
+  kwargs:
+    chunks:
+      x: 2400
+      y: 2400
+  meta:
+    category: Soil parameters
+    resolution: 225 metres
+    resolution_real: 7.5 arc seconds
+    source_author: Awad Mohammed Ali
```
[the remark this answers is not in the document]
Comment (RubenImhoff) [reply data/catalogs/wflow_data.yml:18]: Yes, the plan is to publish this work (Awad is working on that). So once published, we can link to the doi of publication and dataset.

File: data/catalogs/wflow_data.yml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  path: hydromt_staging/ksathorfrac_thesis_awadali/*.tif
```
[the remark this answers is not in the document]
Comment (dalmijn) [reply data/catalogs/wflow_data.yml:19]: True, however I want to first rewrite it in another format. I do not have the rights to that folder yet

[the remark this answers is not in the document]
Comment (hboisgon) [reply data/catalogs/wflow_data.yml]: Okay I have access so let me know when the file is ready to be moved!

**Review by hboisgon — CHANGES_REQUESTED** [review]
Thanks for improving this PR @dalmijn.

There's a lot of stuff changed in the end (not just ksathorfrac) so to keep track, could you update:

- suggestions for setup_ksathorfrac
- test the old setup_riverwidth method locally if it still works
- check if the read/write to pcraster maps still work

About the predefined catalogs, seems like your implementation would work!
But... HydroMT core just had to revert main because the new changes in deltares_data and artifact_data catalog did not go as planned when merged. From what I understand they want to review and potentially really change the predefined data catalog behavior... especially the version in the root (now link to commit hashes or main branch for the latest).
Because they want to change, I wonder if it is wise to add predefined catalogs in this way now in HydroMT-Wflow... The possible advantages are version control but so far we only have one version for all our catalogs and being able to work offline with HydroMT which do not yet work because of reading the predefined catalog file itself.
So I wonder if we should still not do something simpler for now and discuss with core for v1 if they are re-working on predefined catalogs if they could make it easier for model specific predefined catalogs to be defined.
Sorry for reverting, with the new problems in core I think it's maybe too early for us to do this in the end... but well your work won't be lost when we can implement it in a safer way for later or to help discuss this with core :)
What do you think?

On another note, for predefined catalogs, could you still:

- remove the wflow_artifact_data and the global ksathorfrac map
- document the predefined catalogs so that users know their name and the data they contain? (especially for wflow_deltares_data)

File: docs/changelog.rst [hunk 7]
```diff
@@ -11,6 +11,8 @@ Unreleased
 
 Added
 -----
+- Catalog specifically for HydroMT-Wflow
+- Setup method for the **KsatHorFrac** parameter
```
[the remark this answers is not in the document]
Comment (hboisgon) [reply docs/changelog.rst]: Still relevant actually

File: hydromt_wflow/toml/wflow/wflow_sbm.toml
[the remark this answers is not in the document]
Comment (dalmijn) [reply hydromt_wflow/toml/wflow/wflow_sbm.toml:1]: They are in the parameter data, but also now shipped again with hydromt_wflow source code

**Comment by dalmijn** [comment]
@hboisgon , this PR became a bit of testing ground for wflow specific catalogs. Although I understand your comments regarding the catalogs and accompanying data, I disagree with some of them to point that I choose to discontinue this PR until we come to a combined discussed design for this (also waiting then for the changes in the core I guess). The pure ksathorfrac setup can be found in this PR #255.

**Pull request closed without merging**